netty/src/main/java/io/grpc/netty/NettyClientStream.java

```diff
@@ -105,113 +115,133 @@ public void start(ClientStreamListener listener) {
     }
-  }
 
-  void transportDataReceived(ByteBuf frame, boolean endOfStream) {
-    transportDataReceived(new NettyReadableBuffer(frame.retain()), endOfStream);
-  }
+    /**
+     * Gets the underlying Netty {@link Http2Stream} for this stream.
+     */
+    @Nullable
+    public Http2Stream http2Stream() {
+      return http2Stream;
+    }
 
-  @Override
-  protected void sendCancel(Status reason) {
-    // Send the cancel command to the handler.
-    writeQueue.enqueue(new CancelClientStreamCommand(this, reason), true);
-  }
+    /**
+     * Intended to be overriden by NettyClientTransport, which has more information about failures.
+     * May only be called from event loop.
+     */
+    protected abstract Status statusFromFailedFuture(ChannelFuture f);
 
-  @Override
-  protected void sendFrame(WritableBuffer frame, boolean endOfStream, boolean flush) {
-    ByteBuf bytebuf = frame == null ? EMPTY_BUFFER : ((NettyWritableBuffer) frame).bytebuf();
-    final int numBytes = bytebuf.readableBytes();
-    if (numBytes > 0) {
-      // Add the bytes to outbound flow control.
-      onSendingBytes(numBytes);
-      writeQueue.enqueue(
-          new SendGrpcFrameCommand(this, bytebuf, endOfStream),
-          channel.newPromise().addListener(new ChannelFutureListener() {
-            @Override
-            public void operationComplete(ChannelFuture future) throws Exception {
-              // Remove the bytes from outbound flow control, optionally notifying
-              // the client that they can send more bytes.
-              onSentBytes(numBytes);
-            }
-          }), flush);
-    } else {
-      // The frame is empty and will not impact outbound flow control. Just send it.
-      writeQueue.enqueue(new SendGrpcFrameCommand(this, bytebuf, endOfStream), flush);
+    @Override
+    protected void http2ProcessingFailed(Status status, Metadata trailers) {
+      transportReportStatus(status, false, trailers);
+      handler.getWriteQueue().enqueue(new CancelClientStreamCommand(this, status), true);
     }
-  }
 
-  @Override
-  protected void returnProcessedBytes(int processedBytes) {
-    handler.returnProcessedBytes(http2Stream, processedBytes);
-    writeQueue.scheduleFlush();
+    @Override
+    public void bytesRead(int processedBytes) {
+      handler.returnProcessedBytes(http2Stream, processedBytes);
+      handler.getWriteQueue().scheduleFlush();
+    }
+
+    @Override
+    protected void deframeFailed(Throwable cause) {
+      http2ProcessingFailed(Status.fromThrowable(cause), new Metadata());
+    }
+
+    void transportHeadersReceived(Http2Headers headers, boolean endOfStream) {
+      if (endOfStream) {
+        transportTrailersReceived(Utils.convertTrailers(headers));
+      } else {
+        transportHeadersReceived(Utils.convertHeaders(headers));
+      }
+    }
+
+    void transportDataReceived(ByteBuf frame, boolean endOfStream) {
+      transportDataReceived(new NettyReadableBuffer(frame.retain()), endOfStream);
```

Comment:
No, it is required because the caller isn't donating their reference to us.

This is called from our `Http2FrameAdapter.onDataRead`. Two things to note: 1. `ByteToMessageDecoder` (as extended by `Http2ConnectionHandler`) auto-releases buffers after it calls into the handler (which is... confusing, but life) and 2. since `onDataRead` may not be a noop (the default), `Http2ConnectionHandler` _must_ release the reference after the method returns.